utils/swift_build_support/swift_build_support/products/product_builder.py

```diff
@@ -0,0 +1,61 @@
+# swift_build_support/product_builders/product_builder.py -------*- python -*-
+#
+# This source file is part of the Swift.org open source project
+#
+# Copyright (c) 2014 - 2019 Apple Inc. and the Swift project authors
+# Licensed under Apache License v2.0 with Runtime Library Exception
+#
+# See https://swift.org/LICENSE.txt for license information
+# See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+#
+# ----------------------------------------------------------------------------
+
+import abc
+
+
+class ProductBuilder(object):
+    """
+    Abstract base class for all ProductBuilders.
+
+    An specific ProductBuilder will implement the interface methods depending
+    how the product want to be build. Multiple products can use the same
+    product builder if parametrized right (for example all the products build
+    using CMake).
+
+    Ideally a ProductBuilder will be initialized with references to the
+    invocation arguments, the calculated toolchain, the calculated workspace,
+    and the target host, but the base class doesn't impose those requirements
+    in order to be flexible.
+
+    NOTE: Python doesn't need an explicit abstract base class, but it helps
+    documenting the interface.
+    """
+
+    @abc.abstractmethod
+    def do_build(self):
```

Comment:
I'm not sure if I like the names `do_build`, `do_test` and `do_install`. I think I'd prefer something closer to `run_build` or maybe just `build`. Mostly a stylistic preference and certainly a nit-pick.

Reply:
I can change to whatever people prefer. The current names are the names already used by the Ninja, Indexstore-DB and Sourcekit-LSP (except `do_install`, which I just created by following the patter). That's why I used them. If people prefer other names, I can change those other first, and then change this PR.

Reply:
I would definitely be in favour of naming these without the `do_` prefix.  I don't really care about the ordering - before or after is at your preference @drodriguez .